_source/_posts/2022-06-01-jhipster-digital-ocean.md

```diff
@@ -0,0 +1,411 @@
+
+The next step is to generate the Kubernetes deployment descriptors. In the project root folder, create a `k8s` directory and run the `k8s` JHipster sub-generator:
+
+```bash
+mkdir k8s
+cd k8s
+jhipster k8s
+```
+
+Choose the following options when prompted:
+
+- Type of application: **Microservice application**
+- Root directory: **../**
+- Which applications? (select all)
+- Set up monitoring? **No**
+- Which applications with clustered databases? select **store**
+- Admin password for JHipster Registry: (generate one)
+- Kubernetes namespace: **demo**
+- Docker repository name: (your docker hub username)
+- Command to push Docker image: `docker push`
+- Enable Istio? **No**
+- Kubernetes service type? **LoadBalancer**
+- Use dynamic storage provisioning? **Yes**
+- Use a specific storage class? (leave empty)
+
+**NOTE**: You can leave the Docker repository name blank for running Kubernetes locally, but a repository will be required for the cloud deployment, so go ahead and create a [Docker Hub](https://hub.docker.com/) personal account, and the image pull configuration will be ready for both local and cloud deployments.
+
+Build the `gateway`, `store` and `blog` services container images with Jib. For example, for the `gateway` service:
+
+```bash
+cd ../gateway
+./gradlew bootJar -Pprod jib -Djib.to.image=<docker-repo-name>/gateway
+```
+
+Check the images were uploaded to [DockerHub](https://hub.docker.com), and navigate to the project root folder in the terminal for the next step.
+
+## Configure authentication with OpenID Connect and Okta
+
+One more configuration step before running the architecture locally, let's configure Okta for authentication.
+
+{% include setup/cli.md type="jhipster" %}
+
+The settings from the generated `.okta.env` must be added to the `k8s/registry-k8s/application-configmap.yml`:
+
+```yml
+data:
+  application.yml: |-
+    ...
+    spring:
+      security:
+        oauth2:
+          client:
+            provider:
+              oidc:
+                issuer-uri: https://<your-okta-domain>/oauth2/default
+            registration:
+              oidc:
+                client-id: <client-id>
+                client-secret: <client-secret>
+```
+
+Enable the OIDC authentication in the `jhipster-registry` service by adding the `oauth2` profile in the `k8s/registry-k8s/jhipster-registry.yml` file:
+
+```yml
+- name: SPRING_PROFILES_ACTIVE
+  value: prod,k8s,oauth2
+```
+
+## Run locally with Minikube
+
+Install [Minikube](https://minikube.sigs.k8s.io/docs/start/) and [`kubectl`](https://kubernetes.io/docs/tasks/tools/).
+
+For Minkube, you will need at least 2 CPUs. Start MiniKube with your number of CPUs:
+
+```bash
+cd k8s
+minikube --cpus <ncpu> start
```

Comment:
I got these warnings on startup:

```
    ▪ You are trying to run amd64 binary on M1 system. Please consider running darwin/arm64 binary instead (Download at https://github.com/kubernetes/minikube/releases/download/v1.22.0/minikube-darwin-arm64.)
😄  minikube v1.22.0 on Darwin 12.3.1
🎉  minikube 1.25.2 is available! Download it: https://github.com/kubernetes/minikube/releases/tag/v1.25.2
```

I downloaded from `https://github.com/kubernetes/minikube/releases/download/v1.22.0/minikube-darwin-arm64` and ran the following to fix:

```
chmod +x minikube-darwin-arm64
mv minikube-darwin-arm64 /usr/local/bin/minikube
```

Reply:
Looks much better now:

```
minikube --cpus 2 start
😄  minikube v1.22.0 on Darwin 12.3.1 (arm64)
✨  Using the docker driver based on existing profile
❗  You cannot change the CPUs for an existing minikube cluster. Please first delete the cluster.
👍  Starting control plane node minikube in cluster minikube
🚜  Pulling base image ...
🤷  docker "minikube" container is missing, will recreate.
🔥  Creating docker container (CPUs=8, Memory=16300MB)
```

Reply:
On my Intel MacBook, I see:

```
😄  minikube v1.25.2 on Darwin 12.3.1
🆕  Kubernetes 1.23.3 is now available. If you would like to upgrade, specify: --kubernetes-version=v1.23.3
✨  Using the docker driver based on existing profile
❗  You cannot change the CPUs for an existing minikube cluster. Please first delete the cluster.
👍  Starting control plane node minikube in cluster minikube
🚜  Pulling base image ...
💾  Downloading Kubernetes v1.21.2 preload ...
    > preloaded-images-k8s-v17-v1...: 499.05 MiB / 499.05 MiB  100.00% 3.86 MiB
    > gcr.io/k8s-minikube/kicbase: 361.09 MiB / 361.09 MiB  100.00% 2.77 MiB p/
🤷  docker "minikube" container is missing, will recreate.
🔥  Creating docker container (CPUs=8, Memory=16300MB) ...
🐳  Preparing Kubernetes v1.21.2 on Docker 20.10.7 ...-
```

Reply:
Hi Matt, I don't understand. Should we change the link to minikube? Point to some specific download? I believe the link https://minikube.sigs.k8s.io/docs/start/ should detect the os and architecture.

Reply:
I tried again today with `minikube --cpus=max start` and this time everything starts but `store-mongodb`. 

<img width="1442" alt="Screen Shot 2022-05-17 at 18 40 51" src="https://user-images.githubusercontent.com/17892/168886648-72e3550c-297f-4de4-92db-9c932a1c1542.png">

Maybe it's an issue with the M1. Unfortunately, I won't have my Intel Mac until Thursday. I'm traveling in Iceland this week.

Reply:
It seems your M1 OS ran out of threads. Don't panic, look at the beautiful Blue Lagoon!